File: android-app/src/main/kotlin/com/jworks/kanjiquest/android/ui/settings/SettingsViewModel.kt
package com.jworks.kanjiquest.android.ui.settings

import android.content.Context
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.jworks.kanjiquest.core.domain.UserSessionProvider
import com.jworks.kanjiquest.core.domain.model.LevelProgression
import com.jworks.kanjiquest.core.domain.model.UserLevel
import com.jworks.kanjiquest.core.domain.repository.DevChatRepository
import com.jworks.kanjiquest.core.domain.repository.UserRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class SettingsUiState(
    val soundEnabled: Boolean = true,
    val musicEnabled: Boolean = true,
    val vibrationsEnabled: Boolean = true,
    val notificationsEnabled: Boolean = true,
    val dailyGoal: Int = 20,
    val sessionLength: Int = 10,
    val difficulty: DifficultyLevel = DifficultyLevel.MEDIUM,
    val autoPlayAudio: Boolean = true,
    val showHints: Boolean = true,
    val theme: AppTheme = AppTheme.SYSTEM,
    val isAdmin: Boolean = false,
    val isDeveloper: Boolean = false,
    val effectiveLevel: UserLevel = UserLevel.FREE,
    val adminOverrideLevel: UserLevel? = null,
    val adminPlayerLevelOverride: Int? = null,
    val currentPlayerLevel: Int = 1
)

enum class DifficultyLevel(val displayName: String) {
    EASY("Easy"),
    MEDIUM("Medium"),
    HARD("Hard")
}

enum class AppTheme(val displayName: String) {
    LIGHT("Light"),
    DARK("Dark"),
    SYSTEM("System Default")
}

@HiltViewModel
class SettingsViewModel @Inject constructor(
    @ApplicationContext private val context: Context,
    private val userRepository: UserRepository,
    private val userSessionProvider: UserSessionProvider,
    private val devChatRepository: DevChatRepository
) : ViewModel() {

    private val prefs = context.getSharedPreferences("kanjiquest_settings", Context.MODE_PRIVATE)

    private val _uiState = MutableStateFlow(SettingsUiState())
    val uiState: StateFlow<SettingsUiState> = _uiState.asStateFlow()

    init {
        loadSettings()
    }

    private fun loadSettings() {
        viewModelScope.launch {
            val profile = try { userRepository.getProfile() } catch (_: Exception) { null }
            val email = userSessionProvider.getUserEmail()
            val isDev = if (email != null) {
                try { devChatRepository.isDeveloper(email) } catch (_: Exception) { false }
            } else false
            _uiState.value = SettingsUiState(
                soundEnabled = prefs.getBoolean("sound_enabled", true),
                musicEnabled = prefs.getBoolean("music_enabled", true),
                vibrationsEnabled = prefs.getBoolean("vibrations_enabled", true),
                notificationsEnabled = prefs.getBoolean("notifications_enabled", true),
                dailyGoal = prefs.getInt("daily_goal", 20),
                sessionLength = prefs.getInt("session_length", 10),
                difficulty = DifficultyLevel.valueOf(
                    prefs.getString("difficulty", DifficultyLevel.MEDIUM.name) ?: DifficultyLevel.MEDIUM.name
                ),
                autoPlayAudio = prefs.getBoolean("auto_play_audio", true),
                showHints = prefs.getBoolean("show_hints", true),
                theme = AppTheme.valueOf(
                    prefs.getString("theme", AppTheme.SYSTEM.name) ?: AppTheme.SYSTEM.name
                ),
                isAdmin = userSessionProvider.isAdmin(),
                isDeveloper = isDev,
                effectiveLevel = userSessionProvider.getEffectiveLevel(),
                adminOverrideLevel = userSessionProvider.getAdminOverrideLevel(),
                adminPlayerLevelOverride = userSessionProvider.getAdminPlayerLevelOverride(),
                currentPlayerLevel = profile?.level ?: 1
            )
        }
    }

    fun setAdminOverrideLevel(level: UserLevel?) {
        userSessionProvider.setAdminOverrideLevel(level)
        _uiState.value = _uiState.value.copy(
            adminOverrideLevel = level,
            effectiveLevel = userSessionProvider.getEffectiveLevel()
        )
    }

    fun setAdminPlayerLevelOverride(level: Int?) {
        userSessionProvider.setAdminPlayerLevelOverride(level)
        _uiState.value = _uiState.value.copy(adminPlayerLevelOverride = level)
    }

    fun toggleSound() {
        val newValue = !_uiState.value.soundEnabled
        prefs.edit().putBoolean("sound_enabled", newValue).apply()
        _uiState.value = _uiState.value.copy(soundEnabled = newValue)
    }

    fun toggleMusic() {
        val newValue = !_uiState.value.musicEnabled
        prefs.edit().putBoolean("music_enabled", newValue).apply()
        _uiState.value = _uiState.value.copy(musicEnabled = newValue)
    }

    fun toggleVibrations() {
        val newValue = !_uiState.value.vibrationsEnabled
        prefs.edit().putBoolean("vibrations_enabled", newValue).apply()
        _uiState.value = _uiState.value.copy(vibrationsEnabled = newValue)
    }

    fun toggleNotifications() {
        val newValue = !_uiState.value.notificationsEnabled
        prefs.edit().putBoolean("notifications_enabled", newValue).apply()
        _uiState.value = _uiState.value.copy(notificationsEnabled = newValue)
    }

    fun toggleAutoPlayAudio() {
        val newValue = !_uiState.value.autoPlayAudio
        prefs.edit().putBoolean("auto_play_audio", newValue).apply()
        _uiState.value = _uiState.value.copy(autoPlayAudio = newValue)
    }

    fun toggleShowHints() {
        val newValue = !_uiState.value.showHints
        prefs.edit().putBoolean("show_hints", newValue).apply()
        _uiState.value = _uiState.value.copy(showHints = newValue)
    }

    fun setDailyGoal(goal: Int) {
        prefs.edit().putInt("daily_goal", goal).apply()
        _uiState.value = _uiState.value.copy(dailyGoal = goal)

        viewModelScope.launch {
            try {
                userRepository.updateDailyGoal(goal)
            } catch (e: Exception) {
                // Handle error silently, SharedPreferences still saved
            }
        }
    }

    fun setSessionLength(length: Int) {
        prefs.edit().putInt("session_length", length).apply()
        _uiState.value = _uiState.value.copy(sessionLength = length)
    }

    fun setDifficulty(difficulty: DifficultyLevel) {
        prefs.edit().putString("difficulty", difficulty.name).apply()
        _uiState.value = _uiState.value.copy(difficulty = difficulty)
    }

    fun setTheme(theme: AppTheme) {
        prefs.edit().putString("theme", theme.name).apply()
        _uiState.value = _uiState.value.copy(theme = theme)
    }

    fun resetToDefaults() {
        prefs.edit().clear().apply()
        loadSettings()
    }
}

File: ios-app/KanjiQuest/Screens/Settings/SettingsViewModel.swift
import Foundation
import SharedCore

enum DifficultyLevel: String, CaseIterable {
    case easy = "Easy"
    case medium = "Medium"
    case hard = "Hard"
}

enum AppTheme: String, CaseIterable {
    case light = "Light"
    case dark = "Dark"
    case system = "System Default"
}

/// Settings view model. Mirrors Android's SettingsViewModel.kt.
@MainActor
class SettingsViewModel: ObservableObject {
    @Published var soundEnabled = true
    @Published var musicEnabled = true
    @Published var vibrationsEnabled = true
    @Published var notificationsEnabled = true
    @Published var dailyGoal = 20
    @Published var sessionLength = 10
    @Published var difficulty: DifficultyLevel = .medium
    @Published var autoPlayAudio = true
    @Published var showHints = true
    @Published var theme: AppTheme = .system
    @Published var isAdmin = false
    @Published var isDeveloper = false
    @Published var currentPlayerLevel = 1
    @Published var adminPlayerLevelOverride: Int?

    private let defaults = UserDefaults.standard

    func load(container: AppContainer) {
        soundEnabled = defaults.object(forKey: "sound_enabled") as? Bool ?? true
        musicEnabled = defaults.object(forKey: "music_enabled") as? Bool ?? true
        vibrationsEnabled = defaults.object(forKey: "vibrations_enabled") as? Bool ?? true
        notificationsEnabled = defaults.object(forKey: "notifications_enabled") as? Bool ?? true
        dailyGoal = defaults.integer(forKey: "daily_goal").nonZero ?? 20
        sessionLength = defaults.integer(forKey: "session_length").nonZero ?? 10
        autoPlayAudio = defaults.object(forKey: "auto_play_audio") as? Bool ?? true
        showHints = defaults.object(forKey: "show_hints") as? Bool ?? true

        if let diffStr = defaults.string(forKey: "difficulty"),
           let diff = DifficultyLevel(rawValue: diffStr) {
            difficulty = diff
        }
        if let themeStr = defaults.string(forKey: "theme"),
           let t = AppTheme(rawValue: themeStr) {
            theme = t
        }

        Task {
            let profile = try? await container.userRepository.getProfile()
            currentPlayerLevel = Int(profile?.level ?? 1)
        }
    }

    func toggleSound() {
        soundEnabled.toggle()
        defaults.set(soundEnabled, forKey: "sound_enabled")
    }

    func toggleMusic() {
        musicEnabled.toggle()
        defaults.set(musicEnabled, forKey: "music_enabled")
    }

    func toggleVibrations() {
        vibrationsEnabled.toggle()
        defaults.set(vibrationsEnabled, forKey: "vibrations_enabled")
    }

    func toggleNotifications() {
        notificationsEnabled.toggle()
        defaults.set(notificationsEnabled, forKey: "notifications_enabled")
    }

    func toggleAutoPlayAudio() {
        autoPlayAudio.toggle()
        defaults.set(autoPlayAudio, forKey: "auto_play_audio")
    }

    func toggleShowHints() {
        showHints.toggle()
        defaults.set(showHints, forKey: "show_hints")
    }

    func setDailyGoal(_ goal: Int) {
        dailyGoal = goal
        defaults.set(goal, forKey: "daily_goal")
    }

    func setSessionLength(_ length: Int) {
        sessionLength = length
        defaults.set(length, forKey: "session_length")
    }

    func setDifficulty(_ diff: DifficultyLevel) {
        difficulty = diff
        defaults.set(diff.rawValue, forKey: "difficulty")
    }

    func setTheme(_ t: AppTheme) {
        theme = t
        defaults.set(t.rawValue, forKey: "theme")
    }

    func resetToDefaults() {
        let keys = ["sound_enabled", "music_enabled", "vibrations_enabled", "notifications_enabled",
                    "daily_goal", "session_length", "difficulty", "auto_play_audio", "show_hints", "theme"]
        keys.forEach { defaults.removeObject(forKey: $0) }
        soundEnabled = true; musicEnabled = true; vibrationsEnabled = true; notificationsEnabled = true
        dailyGoal = 20; sessionLength = 10; difficulty = .medium; autoPlayAudio = true
        showHints = true; theme = .system
    }
}

private extension Int {
    var nonZero: Int? { self == 0 ? nil : self }
}
